### PR TITLE
Change emulator repository to store file per key

### PR DESCRIPTION
**Pull request by giladsu**

### CODE REVIEW

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 1]
```diff
@@ -148,40 +158,38 @@ public Task<T> GetStateAsync<T>(string key, CancellationToken cancellationToken)
...
         {
-            lock (RepositoryFileLock)
-            {
-                using (StreamWriter writer = File.CreateText(this.repositoryFilePath))
-                {
-                    JsonSerializer serializer = new JsonSerializer();
-                    serializer.Serialize(writer, this.stateRepository);
-                }
-            }
+            return key.Hash();
```
Comment (giladsu) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: A little worried about collisions here, but IIUC this is what we are doing in the actual blob storage....

File: test/SDK/MonitoringApplianceEmulatorTests/2c70e12b7a0646f92279f427c7b38e7334d8e5389cff167a1dc30e73f826b683 [hunk 2]
```diff
@@ -0,0 +1 @@
+{"Field1":"testdata","Field2":["2018-02-15T00:00:00"],"Field3":true,"Field4":null}
```
Comment (liadtal) [inline test/SDK/MonitoringApplianceEmulatorTests/2c70e12b7a0646f92279f427c7b38e7334d8e5389cff167a1dc30e73f826b683]: Was this file added intentionally?

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 3]
```diff
@@ -28,10 +28,12 @@ public class EmulationStateRepository : IStateRepository
         private static readonly object RepositoryFileLock = new object();
 
         private readonly ConcurrentDictionary<string, string> stateRepository = new ConcurrentDictionary<string, string>();
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >stateRepository [](start = 62, length = 15)

I suggest removing this. This is a memory copy of the state - we don't really need it, and it makes the class more complicated than it should be.

**Comment by liadtal** [comment]
>         private ConcurrentDictionary<string, string> ReadSavedRepository()

If you decide to keep the stateRepository member, this method needs to be fixed (I suggest removing the object and the method)

---
Refers to: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs:238 in 31a3003. [](commit_id = 31a300392c1e7da43ea61d8194b98cefa2fb25fd, deletion_comment = False)

**Comment by liadtal** [comment]
>         private string GenerateRepositoryPath()

Just generate the folder path - not the file (we don't use the big json file anymore)

---
Refers to: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs:216 in 31a3003. [](commit_id = 31a300392c1e7da43ea61d8194b98cefa2fb25fd, deletion_comment = False)

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 4]
```diff
@@ -104,9 +108,12 @@ public Task StoreStateAsync<T>(string key, T state, CancellationToken cancellati
                 throw new StateTooBigException(serializedState.Length, MaxSerializedStateLength);
             }
 
-            this.stateRepository[key] = serializedState;
-
-            this.Save();
+            var lockObj = this.lockRepository.GetOrAdd(key, k => new object());
+            lock (lockObj)
+            {
+                File.WriteAllText(Path.Combine(this.repositoryDir, this.GetFilename(key)), serializedState);
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >Path.Combine(this.repositoryDir, this.GetFilename(key)) [](start = 34, length = 55)

You can make the Path.Combine part of GetFilename - will save a few code duplications in this class

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 5]
```diff
@@ -28,10 +28,12 @@ public class EmulationStateRepository : IStateRepository
         private static readonly object RepositoryFileLock = new object();
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >RepositoryFileLock [](start = 39, length = 18)

If you remove the stateRepository object, you can remove this as well

**Comment by liadtal** [comment]
>         public EmulationStateRepository(string smartDetectorId)

Would be nicer if this ctor were to call the other ctor instead of duplicating the code

---
Refers to: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs:44 in 31a3003. [](commit_id = 31a300392c1e7da43ea61d8194b98cefa2fb25fd, deletion_comment = False)

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 6]
```diff
@@ -28,10 +28,12 @@ public class EmulationStateRepository : IStateRepository
...
 
         private readonly ConcurrentDictionary<string, string> stateRepository = new ConcurrentDictionary<string, string>();
+        private readonly ConcurrentDictionary<string, object> lockRepository = new ConcurrentDictionary<string, object>();
 
         private readonly string smartDetectorIdentifier;
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >private readonly string smartDetectorIdentifier; [](start = 8, length = 48)

This is only used in the ctor - no need to keep this as a member

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 7]
```diff
@@ -28,10 +28,12 @@ public class EmulationStateRepository : IStateRepository
...
 
         private readonly string repositoryFilePath;
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >repositoryFilePath [](start = 32, length = 18)

Should also be removed

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 8]
```diff
@@ -104,9 +108,12 @@ public Task StoreStateAsync<T>(string key, T state, CancellationToken cancellati
                 throw new StateTooBigException(serializedState.Length, MaxSerializedStateLength);
             }
 
-            this.stateRepository[key] = serializedState;
-
-            this.Save();
+            var lockObj = this.lockRepository.GetOrAdd(key, k => new object());
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >this.lockRepository.GetOrAdd(key, k => new object()) [](start = 26, length = 52)

this is an overkill - in the emulator we don't expect a lot of concurrency here
just keep a single lock for all files

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj [hunk 9]
```diff
@@ -22,7 +22,7 @@
   <Import Project="$(NuProjPath)\NuProj.props" Condition="Exists('$(NuProjPath)\NuProj.props')" />
   <PropertyGroup Label="Configuration">
     <Id>Microsoft.Azure.Monitoring.SmartDetectors</Id>
-    <Version>0.7.0.0</Version>
+    <Version>0.8.0.0</Version>
```
Comment (EranGon) [inline src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: >8 [](start = 15, length = 1)

please use 0.7.1.0

**Comment by EranGon** [comment]
>         public EmulationStateRepository(string smartDetectorId, string statesContainerFilePath)

who uses this ctor? please remove if not needed

---
Refers to: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs:59 in 31a3003. [](commit_id = 31a300392c1e7da43ea61d8194b98cefa2fb25fd, deletion_comment = False)

**Comment by EranGon** [comment]
>         public EmulationStateRepository(string smartDetectorId, string statesContainerFilePath)

there is no container here - change to repositoryFilePath

---
Refers to: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs:59 in 31a3003. [](commit_id = 31a300392c1e7da43ea61d8194b98cefa2fb25fd, deletion_comment = False)

**Comment by EranGon** [comment]
>             }

you'll need to delete the whole folder here

---
Refers to: test/SDK/MonitoringApplianceEmulatorTests/EmulationStateRepositoryTests.cs:33 in 31a3003. [](commit_id = 31a300392c1e7da43ea61d8194b98cefa2fb25fd, deletion_comment = False)

File: test/SDK/MonitoringApplianceEmulatorTests/2c70e12b7a0646f92279f427c7b38e7334d8e5389cff167a1dc30e73f826b683 (hunk 2 above)
Comment (giladsu) [reply test/SDK/MonitoringApplianceEmulatorTests/2c70e12b7a0646f92279f427c7b38e7334d8e5389cff167a1dc30e73f826b683]: Weird! 

---
In reply to: [223213890](https://github.com/liadtal/Smart-Alerts/pull/219#discussion_r223213890) [](ancestors = 223213890)

**Comment by giladsu** [comment]
>         public EmulationStateRepository(string smartDetectorId, string statesContainerFilePath)

See comments - it is used in UTs.

---
In reply to: [427791572](https://github.com/liadtal/Smart-Alerts/pull/219#issuecomment-427791572) [](ancestors = 427791572)

---
Refers to: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs:59 in 31a3003. [](commit_id = 31a300392c1e7da43ea61d8194b98cefa2fb25fd, deletion_comment = False)

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs (hunk 4 above)
Comment (giladsu) [reply src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: Love it.

---
In reply to: [223214159](https://github.com/liadtal/Smart-Alerts/pull/219#discussion_r223214159) [](ancestors = 223214159)

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs (hunk 8 above)
Comment (giladsu) [reply src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: as discussed f2f/

---
In reply to: [223283800](https://github.com/liadtal/Smart-Alerts/pull/219#discussion_r223283800) [](ancestors = 223283800)

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj (hunk 9 above)
Comment (giladsu) [reply src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: I can't - the 0.8 version has a fix for a blocking bug for us.

---
In reply to: [223320767](https://github.com/liadtal/Smart-Alerts/pull/219#discussion_r223320767) [](ancestors = 223320767)

File: test/SDK/MonitoringApplianceEmulatorTests/MonitoringApplianceEmulatorTests.csproj [hunk 10]
```diff
@@ -4,6 +4,9 @@
 	<StyleCopTreatErrorsAsWarnings>False</StyleCopTreatErrorsAsWarnings>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <PlatformTarget>x64</PlatformTarget>
+  </PropertyGroup>
```
Comment (giladsu) [inline test/SDK/MonitoringApplianceEmulatorTests/MonitoringApplianceEmulatorTests.csproj]: Reverted.

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 11]
```diff
@@ -148,20 +142,16 @@ public Task<T> GetStateAsync<T>(string key, CancellationToken cancellationToken)
...
+            var lockObj = LockRepository.GetOrAdd(key, k => new object());
+            lock (lockObj)
             {
-                T state = default(T);
-                try
-                {
-                    state = JsonConvert.DeserializeObject<T>(value);
-                }
-                catch (Exception ex)
+                var statePath = this.GetStateFilePath(key);
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >var statePath = this.GetStateFilePath(key); [](start = 16, length = 43)

you can move this to outside of the lock block

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 12]
```diff
@@ -104,9 +94,11 @@ public Task StoreStateAsync<T>(string key, T state, CancellationToken cancellati
                 throw new StateTooBigException(serializedState.Length, MaxSerializedStateLength);
             }
 
-            this.stateRepository[key] = serializedState;
-
-            this.Save();
+            var lockObj = LockRepository.GetOrAdd(key, k => new object());
+            lock (lockObj)
+            {
+                File.WriteAllText(this.GetStateFilePath(key), serializedState);
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >WriteAllText [](start = 21, length = 12)

please change to WriteAllTextAsync and make this method async

File: src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs [hunk 13]
```diff
@@ -148,20 +142,16 @@ public Task<T> GetStateAsync<T>(string key, CancellationToken cancellationToken)
...
-                try
-                {
-                    state = JsonConvert.DeserializeObject<T>(value);
-                }
-                catch (Exception ex)
+                var statePath = this.GetStateFilePath(key);
+                if (File.Exists(statePath))
                 {
-                    throw new StateSerializationException(ex);
+                    var serializedState = File.ReadAllText(statePath);
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/State/EmulationStateRepository.cs]: >ReadAllText [](start = 47, length = 11)

ReadAllTextAsync

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj (hunk 9 above)
Comment (EranGon) [reply src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: what? 0.8 version does not exist...

---
In reply to: [223595752](https://github.com/liadtal/Smart-Alerts/pull/219#discussion_r223595752) [](ancestors = 223595752,223320767)

**Pull request merged**